libs/wasm-detective/src/detective.rs

```diff
@@ -217,42 +267,62 @@ fn validate_request(request: &Request) -> Result<(), CustomError> {
 
     Ok(())
 }
-fn recurse_field(request: &Request, val: gjson::Value, f: MatcherFunc) -> bool {
+fn recurse_field(
+    request: &Request,
+    val: gjson::Value,
+    f: MatcherFunc,
+    path: Vec<String>,
+) -> Vec<DetectiveStepResultMatch> {
+    let mut res: Vec<DetectiveStepResultMatch> = Vec::new();
+
     match val.kind() {
         gjson::Kind::String | gjson::Kind::Number | gjson::Kind::True | gjson::Kind::False => {
-            if let Ok(res) = f(request, val) {
-                return res;
+            // Avoiding borrow, gjson::Value does not support cloning
+            let v = gjson::parse(val.json());
+
+            if let Ok(found) = f(request, v) {
+                if found {
+                    let result = DetectiveStepResultMatch {
+                        type_: ::protobuf::EnumOrUnknown::new(request.match_type),
+                        path: path.join("."),
+                        value: val.str().to_owned().into_bytes(),
+                        special_fields: Default::default(),
+                    };
+
+                    res.push(result);
+                }
             }
         }
 
         gjson::Kind::Object => {
-            let mut found: bool = false;
-            val.each(|_, value| {
-                if recurse_field(request, value, f) {
-                    found = true;
+            val.each(|key, value| {
+                let mut cur_path = path.clone();
+                cur_path.push(key.to_string());
+
+                let matches = recurse_field(request, value, f, cur_path.clone());
+                if !matches.is_empty() {
+                    res.extend(matches);
                 }
                 true
             });
-
-            if found {
-                return true;
-            }
         }
         gjson::Kind::Array => {
-            let mut found: bool = false;
+            let mut cur_path = path.clone();
+            cur_path.push("#".to_string());
+
             val.each(|_, value| {
-                if recurse_field(request,value, f) {
-                    found = true;
+                let matches = recurse_field(request, value, f, cur_path.clone());
+                if !matches.is_empty() {
+                    res.extend(matches);
```

Comment:
Went down a rabbit hole here because I didn't catch the negated `is_empty()` and found myself reading a bikeshedding discussion between Rust programmers talking about visibility of `!is_empty()`... 

This is fine.